[full-ci] Resharing

kobergj opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the oCIS component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next version of oCIS.

Please set the following labels:

- Set label "Status:Needs-Review" for review or "Status:In-Progress" if the PR still has open tasks.
- Assignment: assign to self
- Reviewers: pick at least one
-->

## Description
Adds and activates resharing functionality for ocis

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- Fixes <issue_link>

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- test environment:
- test case 1:
- test case 2:
- ...

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in owncloud.github.io/ -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

kobergj commented on tests/acceptance/features/apiSpaces/resharing.feature:
```diff
@@ -0,0 +1,85 @@
+      | Carol | 100  | should be     |
+
+  Scenario Outline: Deleting reshares
+    When user "Gina" has been created with default attributes and without skeleton files    
+    And user "Carol" has shared folder "Shares/folder" with user "Gina" with permissions "17"
+    And user "Gina" accepts share "/folder" offered by user "Carol" using the sharing API
+    And as "Gina" folder "Shares/folder" should exist
+    Then user "<user>" deletes the last share using the sharing API
+    And the OCS status code should be "<code>"
+    And as "Gina" folder "Shares/folder" <exists>
+    And as "Carol" folder "Shares/folder" should exist
```
@phil-davis this test is failing since the last master rebase. I think it is because of the line
```
Then user "<user>" deletes the last share using the sharing API
```
and the recently merged "last share" improvements.
I believe my testcode is not correct any more (or it never was).
What I am trying to do is: As user "<user>" delete the share to user "Gina"

Can you help me how I can achieve this most easily?

phil-davis replied:
Sorry I took so long! Yes, there was a regression in the core test code for this type of scenario. I could reproduce it in core by creating a similar scenario there locally and running it. PR https://github.com/owncloud/core/pull/40133 should fix it. I will get the core commit id updated tomorrow morning...

phil-davis replied:
@kobergj the core commit id has been bumped in PR https://github.com/owncloud/ocis/pull/3961

Please rebase and these tests should work better.

ScharfViktor commented on tests/acceptance/features/apiSpaces/resharing.feature:
```diff
@@ -0,0 +1,85 @@
+      | Alice    |
+      | Brian    |
+      | Carol    |
+      | Damian   |
```
You can move this "Create User" step here: Amber, Fred, Gina. You can also use one of the users in feature:50 56 71

ScharfViktor commented on tests/acceptance/features/apiSpaces/resharing.feature:
```diff
@@ -0,0 +1,85 @@
+      | Damian   |
+    And user "Alice" has created folder "folder"
+    And user "Alice" has shared folder "folder" with user "Brian" with permissions "31"
+    And user "Brian" accepts share "/folder" offered by user "Alice" using the sharing API
```
better use given step: `And user "Brian" has accepted share "/folder" offered by user "Alice"`

ScharfViktor commented on tests/acceptance/features/apiSpaces/resharing.feature:
```diff
@@ -0,0 +1,85 @@
+    And user "Alice" has created folder "folder"
+    And user "Alice" has shared folder "folder" with user "Brian" with permissions "31"
+    And user "Brian" accepts share "/folder" offered by user "Alice" using the sharing API
+    And as "Brian" folder "Shares/folder" should exist
```
not sure that we need check it in GivenSteps. If this folder is not present, the test will break in the next step. in scenario we can use this checking

ScharfViktor commented on tests/acceptance/features/apiSpaces/resharing.feature:
```diff
@@ -0,0 +1,85 @@
+    And the OCS status message should be "Cannot set the requested share permissions"
+
+  Scenario Outline: Editing reshares
+    When user "Fred" has been created with default attributes and without skeleton files    
+    And user "Carol" has shared folder "Shares/folder" with user "Fred" with permissions "17"
+    And user "Fred" accepts share "/folder" offered by user "Carol" using the sharing API
+    And as "Fred" folder "Shares/folder" should exist
+    Then user "<user>" updates the last share using the sharing API with
+      | permissions | 31 |
+    And the OCS status code should be "<code>"
+    And user "Fred" <canUpload> able to upload file "filesForUpload/textfile.txt" to "/Shares/folder/textfile.txt"
```
creating user you can move to background 

Given section in scenario:

      Given user "Carol" has shared folder "Shares/folder" with user "Fred" with permissions "17"
      And user "Fred" has accepted share "/folder" offered by user "Carol"
      And as "Fred" folder "Shares/folder" should exist

action: 
``` 
When user "<user>" updates the last share using the sharing API with
    | permissions | 31 |
```

assert:
 ```
Then the OCS status code should be "<code>"
And user "Fred" <canUpload> able to upload file "filesForUpload/textfile.txt" to "/Shares/folder/textfile.txt"
```

ScharfViktor commented:
Looks good, great work.

kobergj commented:
@ScharfViktor thanks for reviewing! I (presumably) addressed all issues, could you check again?

Also I added test for file resharing and resharing with different permissions. Please let me know if I should add more testcases.

re-sharing to group will not work atm. I will add a test later

phil-davis commented:
https://drone.owncloud.com/owncloud/ocis/12507/38/6
```
runsh: Total unexpected passed scenarios throughout the test run:
apiSpaces/resharing.feature:80
apiSpaces/resharing.feature:81
```
those can be removed from expected-failures now - good.

phil-davis commented:
https://drone.owncloud.com/owncloud/ocis/12507/47/6
```
runsh: Total unexpected failed scenarios throughout the test run:
apiWebdavEtagPropagation2/copyFileFolder.feature:188
```
Old closed issue https://github.com/owncloud/ocis/issues/3732 - it looks like this Etag test has been a bit flaky in the past. If it fails again we need to chase it up.

kobergj commented:
@phil-davis Thanks! It works smoothly now. 

Regarding `apiWebdavEtagPropagation2/copyFileFolder.feature:188` I had this failing a few times in the past. I will report when it happens again.

fschade reviewed (approved):
LGTM 🚀

micbar reviewed (commented):
LGTM, some small nitpicks

kobergj commented on go.mod:
```diff
@@ -263,3 +263,5 @@ require (
 	stash.kopano.io/kgol/kcc-go/v5 v5.0.1 // indirect
 )
+
+replace github.com/cs3org/go-cs3apis => github.com/kobergj/go-cs3apis v0.0.0-20220610101249-8067918fe8bf
```
I'd love to. I have the PR here: https://github.com/cs3org/cs3apis/pull/172 it just needs approve&merge

butonic replied:
AFAICT this PR does not use the grant Creator... is this line even necessary for this PR? If so can you point me to the place?

kobergj replied:
It wail break the build without the line. It needs the updated cs3apis when using the current reva. I assume that is because go-cs3apis is not an indirect dependency. It is used in several other places. Using different version of it for ocis and reva will confuse the compiler

kobergj commented on extensions/graph/pkg/service/v0/drives.go:
```diff
@@ -498,7 +498,7 @@ func (g Graph) cs3StorageSpaceToDrive(ctx context.Context, baseURL *url.URL, spa
 				identity := libregraph.IdentitySet{User: &libregraph.Identity{Id: &tmp}}
 				switch {
-				case perm.AddGrant:
+				case perm.RemoveGrant:
```
@micbar I added a comment. Please check again if it is clear enough
